### PR TITLE
Save computed y-values in linearization

### PR DESCRIPTION
**Pull request by jlconlin**
@whaeck and I had a conversation where I mentioned that the linearization doesn't preserve the computed y-values—but it does preserve the x-grid, which can then be used to re-compute the y-values. This could be *very* costly, particularly when we would use this capability to linearize resonances. Since calculating a cross section from resonance parameters is expensive, calculating it twice would be (obviously) twice as expensive. 

This pull request stores the computed y-values in a manner similar to how the x-values are stored. It seems like this was unintentionally left out, but I can't guarantee the intent of the original author (@apmccartney).

### CODE REVIEW

**Comment by jlconlin** [comment]
Note that I've also added GitHub Actions to this to ensure that everything compiles and tests pass as we expect. With a small library like this is, GitHub Actions seem to work well.

File: src/twig/linearize/Callable.hpp [hunk 1]
```diff
@@ -1,25 +1,33 @@
-template< typename Xstack, typename Ydata >
+template< typename Xstack, typename Ystack >
 class Callable {
   using Xdata = typename Xstack::value_type;
+  using Ydata = typename Ystack::value_type;
 
-  Xstack* stack;
+public:
```
Comment (whaeck) [inline src/twig/linearize/Callable.hpp]: Why did you make these previously private members public?

I think you should keep these private and add accessor methods to get to the stack if you need it outside the class. By the way, by making these public, you also made bin() and grid() public as far as I can see.

File: src/twig/linearize/Callable/src/bin.hpp [hunk 2]
```diff
@@ -15,9 +15,11 @@ void bin( Xdata xLeft, Xdata xRight,
         x[left], x[right],
         y[left], y[right] ) ){
 
-      this->stack->push_back( x[left] );
+      this->xStack_->push_back( x[left] );
+      this->yStack_->push_back( y[left] );
       if ( not this->xBuffer.size() ){ 
-        this->stack->push_back( x[right] );
+        // this->xStack_->push_back( x[right] );
+        // this->yStack_->push_back( y[right] );
```
Comment (whaeck) [inline src/twig/linearize/Callable/src/bin.hpp]: Why was this commented out? bin() is called in grid() and it appears you're moving this to grid instead?

File: src/twig/linearize/Callable/src/grid.hpp [hunk 3]
```diff
@@ -16,14 +16,19 @@ void grid( ForwardIterator& first,
...
+    yRight = functor( xRight );
     this->bin( xLeft, xRight, yLeft, yRight,
                std::forward<Functor>(functor),
                std::forward<Convergence>(criterion),
                std::forward<Midpoint>(midpoint) );
     xLeft = xRight;
     yLeft = yRight;
   }
+  this->xStack_->push_back( xRight );
+  this->yStack_->push_back( yRight );
```
Comment (whaeck) [inline src/twig/linearize/Callable/src/grid.hpp:33]: When first == last from the get-go, you will be pushing back default initialised values. Are you sure you want to do that?
Using
```cpp
auto xRight = xLeft;
auto yRight = yLeft;
```
may be a better idea to initialise these values.

File: src/twig/linearize/Callable/test/Callable.test.cpp [hunk 4]
```diff
@@ -28,33 +37,31 @@ SCENARIO("Broken stick implementation"){
...
-    const auto trial = 0.5 * ( y( left() ) + y( right() ) );
-    const auto reference = y( midpoint );
-    CHECK( criterion( trial, reference,
-                       x( left() ), x( right() ),
-                       y( left() ), y( right() ) ) );
-    ++iterator;
+
+  auto length = xInstance.size();
+  std::cout << "xlength: " << xInstance.size() << std::endl;
+  std::cout << "ylength: " << yInstance.size() << std::endl;
```
Comment (whaeck) [inline src/twig/linearize/Callable/test/Callable.test.cpp]: Why write this out? You should check whether or not the x and y lengths are the same.

File: src/twig/linearize/Callable/test/Callable.test.cpp [hunk 5]
```diff
@@ -28,33 +37,31 @@ SCENARIO("Broken stick implementation"){
...
   }
   
-  auto iterator = instance.begin();
+  auto iterator = xInstance.begin();
 
   auto left = [&iterator](){ return iterator[0]; };
   auto right = [&iterator](){ return iterator[1]; };
 
   auto x = []( auto value ){ return value; };
   auto y = [&]( auto value ){ return functor( value ); };
```
Comment (whaeck) [inline src/twig/linearize/Callable/test/Callable.test.cpp]: I believe these are no longer used?

**Review by whaeck — CHANGES_REQUESTED** [review]

File: src/twig/linearize/Callable.hpp (hunk 1 above)
Comment (jlconlin) [reply src/twig/linearize/Callable.hpp]: I accidentally left `public:` in that was added in an earlier step of this work.

File: src/twig/linearize/Callable/src/bin.hpp (hunk 2 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/src/bin.hpp]: Originally, the last value was added here. I accidentally added it again in the previous Pull Request. I've now returned it here.

File: src/twig/linearize/Callable/src/grid.hpp (hunk 3 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/src/grid.hpp:33]: Yeah, I'm not doing this here anymore. Thanks for catching this.

File: src/twig/linearize/Callable/test/Callable.test.cpp (hunk 5 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/test/Callable.test.cpp]: probably right.

File: src/twig/linearize/Callable/test/Callable.test.cpp (hunk 4 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/test/Callable.test.cpp]: Debug process. Forgot to remove them.

**Review by jlconlin — COMMENTED** [review]
I believe I've fixed the issues you caught. Thanks for being thorough.

File: src/twig/linearize/Callable/test/Callable.test.cpp [hunk 6]
```diff
@@ -28,33 +37,20 @@ SCENARIO("Broken stick implementation"){
...
-
-  while ( iterator != std::prev( instance.end() ) ){
-    const auto midpoint = 0.5 * ( x( left() ) + x( right() ) );
-    const auto trial = 0.5 * ( y( left() ) + y( right() ) );
-    const auto reference = y( midpoint );
-    CHECK( criterion( trial, reference,
-                       x( left() ), x( right() ),
-                       y( left() ), y( right() ) ) );
-    ++iterator;
+  auto length = xInstance.size();
```
Comment (whaeck) [inline src/twig/linearize/Callable/test/Callable.test.cpp:45]: You should still check that the vectors are the same size after the linearisation. The for loop that follows assumes that there are no additional elements in the y vector but that is not the same as verifying that they are the same length.

File: src/twig.hpp [hunk 7]
```diff
@@ -6,6 +6,7 @@
 #include <numeric>
 #include <vector>
 #include <iostream>
+#include <memory>
```
Comment (whaeck) [inline src/twig.hpp]: Why is this included? I don't see where it is used.

File: src/twig/linearize/Callable.hpp [hunk 8]
```diff
@@ -1,25 +1,32 @@
...
 public:
-  Callable( Xstack& xStack ) : stack( &xStack ){}
-  void xStack( Xstack& stack ){ this->stack = &stack; }
+Callable( Xstack& xStack, Ystack& yStack ) : 
+  xStack_( &xStack ),
+  yStack_( &yStack )
+{}
+
+  void xStack( Xstack& stack ){ this->xStack_ = &stack; }
+  void yStack( Ystack& stack ){ this->yStack_ = &stack; }
```
Comment (whaeck) [inline src/twig/linearize/Callable.hpp]: Are these functions actually used? This essentially would allow you to change the underlying stacks, and that may be a bad idea (what would happen if I only change the x stack?). You can just create a clean callable anyway.

File: src/twig/linearize/Callable/test/Callable.test.cpp [hunk 9]
```diff
@@ -8,9 +8,18 @@
 
 using namespace njoy::twig;
 
+void printv( const std::vector< double >& v, std::string name ){
+  std::cout << name << ": ";
+  for( auto x: v ){
+    std::cout << x << ", ";
+  }
+  std::cout << std::endl;
+}
```
Comment (whaeck) [inline src/twig/linearize/Callable/test/Callable.test.cpp]: this function is unused.

**Review by whaeck — CHANGES_REQUESTED** [review]

File: src/twig/linearize/Callable/src/bin.hpp (hunk 2 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/src/bin.hpp]: Never mind, there is no `push_back` here anymore. This is where the value was added twice.

File: src/twig/linearize/Callable/src/grid.hpp (hunk 3 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/src/grid.hpp:33]: Okay, upon further investigation this needs to stay here. Without these lines, the right most value of the input grid (and associated functor evaluation) does not get added. Without this, other codes downstream (i.e., RECONR) get errors because the arrays don't line up properly.

I'm not sure what to do about the situation where we begin with `first == last` right off the bat. Suggestions?

File: src/twig/linearize/Callable/test/Callable.test.cpp (hunk 5 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/test/Callable.test.cpp]: They are gone.

File: src/twig/linearize/Callable/test/Callable.test.cpp (hunk 6 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/test/Callable.test.cpp:45]: I've added some tests on this.

File: src/twig.hpp (hunk 7 above)
Comment (jlconlin) [reply src/twig.hpp]: Good catch. I included it because I was going to try to use a `unique_ptr`, but couldn't make it work. Forgot tor remove the `#include` statement. It is now gone.

File: src/twig/linearize/Callable.hpp (hunk 8 above)
Comment (jlconlin) [reply src/twig/linearize/Callable.hpp]: Currently there is no test that uses these functions. You would do this if you were going to re-linearize the same functor on a different grid. Seems like a bad idea anyway. I'll remove them.

File: src/twig/linearize/Callable/test/Callable.test.cpp (hunk 9 above)
Comment (jlconlin) [reply src/twig/linearize/Callable/test/Callable.test.cpp]: It's gone.

**Comment by jlconlin** [comment]
I think I've addressed all the comments. They were very helpful.

**Comment by whaeck** [comment]
@jlconlin It would help if you pushed your last changes :-)

**Review by whaeck — APPROVED** [review]
